default/AI/AIstate.py

```diff
@@ -662,6 +661,14 @@ def rate_fleet(self, fleet_id, enemy_stats=None, ship_info=None):
                 shields = ship.currentMeterValue(fo.meterType.shield)
                 stats['structure'] = structure
                 stats['shields'] = shields
+                stats['attack'] = 0
+                stats['attacks'] = {}
+                design = fo.getShipDesign(design_id)
+                for partname in design.parts:
```

Comment:
You need to check if `fo.getShipDesign(design_id)` returned a valid design. I'm getting these errors when testing:
```
2015-12-26 21:10:03.836602 [error] AI : PythonLoggingWrapper.cpp:41 : Traceback (most recent call last):
2015-12-26 21:10:03.836644 [error] AI : PythonLoggingWrapper.cpp:41 :   File "/Users/user/SoftwareProjekte/FO/src/Xcode/build/Test/FreeOrion.app/Contents/Resources/default/AI/freeorion_tools.py", line 92, in wrapper
2015-12-26 21:10:03.836723 [error] AI : PythonLoggingWrapper.cpp:41 :     return function(*args, **kw)
2015-12-26 21:10:03.836758 [error] AI : PythonLoggingWrapper.cpp:41 :   File "/Users/user/SoftwareProjekte/FO/src/Xcode/build/Test/FreeOrion.app/Contents/Resources/default/AI/freeorion_debug/listeners.py", line 42, in wrapper
2015-12-26 21:10:03.836800 [error] AI : PythonLoggingWrapper.cpp:41 :     res = funct(*args)
2015-12-26 21:10:03.836827 [error] AI : PythonLoggingWrapper.cpp:41 :   File "/Users/user/SoftwareProjekte/FO/src/Xcode/build/Test/FreeOrion.app/Contents/Resources/default/AI/FreeOrionAI.py", line 195, in generateOrders
2015-12-26 21:10:03.836920 [error] AI : PythonLoggingWrapper.cpp:41 :     foAIstate.refresh()  # checks exploration border & clears roles/missions of missing fleets & updates fleet locs & threats
2015-12-26 21:10:03.836949 [error] AI : PythonLoggingWrapper.cpp:41 :   File "/Users/user/SoftwareProjekte/FO/src/Xcode/build/Test/FreeOrion.app/Contents/Resources/default/AI/AIstate.py", line 200, in refresh
2015-12-26 21:10:03.837041 [error] AI : PythonLoggingWrapper.cpp:41 :     self.update_system_status()
2015-12-26 21:10:03.837076 [error] AI : PythonLoggingWrapper.cpp:41 :   File "/Users/user/SoftwareProjekte/FO/src/Xcode/build/Test/FreeOrion.app/Contents/Resources/default/AI/AIstate.py", line 326, in update_system_status
2015-12-26 21:10:03.837211 [error] AI : PythonLoggingWrapper.cpp:41 :     e_rating = self.rate_fleet(fleet_id)
2015-12-26 21:10:03.837253 [error] AI : PythonLoggingWrapper.cpp:41 :   File "/Users/user/SoftwareProjekte/FO/src/Xcode/build/Test/FreeOrion.app/Contents/Resources/default/AI/AIstate.py", line 667, in rate_fleet
2015-12-26 21:10:03.837504 [error] AI : PythonLoggingWrapper.cpp:41 :     for partname in design.parts:
2015-12-26 21:10:03.837541 [error] AI : PythonLoggingWrapper.cpp:41 : AttributeError: 'NoneType' object has no attribute 'parts'
```
I guess the case shouldn't actually occur, but it does, so you need to catch it. Of course, tracking down why `fo.getShipDesign(design_id)` returns `None` in certain cases probably needs to be done at some point...